torch/csrc/jit/codegen/cuda/arith.h

```diff
@@ -18,51 +20,251 @@ namespace torch {
 namespace jit {
 namespace fuser {
 
+template <typename Head, typename... Tail>
+struct IsValidArithOpType {
+  static constexpr bool value = !std::is_base_of<Val, Head>::value
```

Comment:
can we use same_as instead of base_of? I could imagine base_of could be useful in the future, but for now I think we only want to compile for Val* and TensorView*, and nothing else.

Reply:
`same_as` is a C++ 20 feature. I assume you refer to `is_same`. 

Changing `is_base_of ` to `is_same` breaks some tests. For example, `testGPU_FusionCodeGen()` has `add(tv0, new Float(2.0))`, so the second operand here is not `Val*` nor `TensorView*`.

Reply:
Does that mean we're instantiating functions for all combinations of arguments?

Reply:
Sorry, I did mean is_same.

Reply:
Yes, all used combinations are instantiated. 

If we want to avoid that, we could have non-template overloads for `binaryOp`, but the motivation of this templated implementation was to avoid the repetitive code. I'm not sure which one would be more preferable in this case.